app/src/main/java/com/google/android/sambadocumentsprovider/browsing/broadcast/BroadcastUtils.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.google.android.sambadocumentsprovider.browsing.broadcast;
+
+import android.util.Log;
+
+import com.google.android.sambadocumentsprovider.browsing.BrowsingException;
+
+import java.net.InetAddress;
+import java.net.InterfaceAddress;
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+
+class BroadcastUtils {
+  private static final int FILE_SERVER_NODE_TYPE = 0x20;
+
+  /**
+   * Generates a NetBIOS name query request.
+   * https://tools.ietf.org/html/rfc1002
+   * Section 4.2.12
+   */
+  static byte[] createPacket() {
+    ByteBuffer os = ByteBuffer.allocate(50);
+
+    char transId = 5566;
+    char broadcastFlag = 0x0010;
+    char questionCount = 1;
+    char answerResourceCount = 0;
+    char authorityResourceCount = 0;
+    char additionalResourceCount = 0;
+
+    os.putChar(transId);
+    os.putChar(broadcastFlag);
+    os.putChar(questionCount);
+    os.putChar(answerResourceCount);
+    os.putChar(authorityResourceCount);
+    os.putChar(additionalResourceCount);
+
+    // Length of name. 16 bytes of name encoded to 32 bytes.
+    os.put((byte) 0x20);
+
+    // '*' character encodes to 2 bytes.
+    os.put((byte) 0x43);
+    os.put((byte) 0x4b);
+
+    // Write the remaining 15 nulls which encode to 30* 0x41
+    for (int i = 0; i < 30; i++) {
+      os.put((byte) 0x41);
+    }
+
+    // Length of next segment.
+    os.put((byte) 0);
+
+    // Question type: Node status
+    os.putChar((char) 0x21);
+
+    // Question class: Internet
+    os.putChar((char) 0x01);
+
+    return os.array();
+  }
+
+  /**
+   * Parses a positive response to NetBIOS name request query.
+   * https://tools.ietf.org/html/rfc1002
+   * Section 4.2.13
+   */
+  static List<String> extractServers(byte[] data) throws BrowsingException {
+    ByteBuffer buffer = ByteBuffer.wrap(data);
+
+    skipBytes(buffer, 2); // Skip transId.
```

Comment:
It looks like I hardcoded transaction id in the JS implementation I wrote (the 5566 is arbitrary) then forgot about it. I'll open a bug to fix my implementation too. But on a quick reread of the spec - the right thing to do would be use a monotonically increasing value each time you send the request, then ignore responses where the transId does not match.

In pratice it probably wont be a problem but potentially another instance? or another application could be sending other requests. I think that's why I didn't skip this value in my implementation. I just never got around to using it.